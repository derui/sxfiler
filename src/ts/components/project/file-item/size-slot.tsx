import * as React from "react";
import bigInt from "big-integer";

import { styled } from "@/components/theme";

enum SizeUnit {
  Byte = "byte",
  KByte = "kbyte",
  MByte = "mbyte",
  GByte = "gbyte",
  TByte = "tbyte",
  Unknown = "unknown",
}

interface SizeData {
  sizeUnit: SizeUnit;
  size: bigInt.BigInteger;
  alignedSize: number;
}

function nextUnit(sizeUnit: SizeUnit): SizeUnit | null {
  switch (sizeUnit) {
    case "byte":
      return SizeUnit.KByte;
    case "kbyte":
      return SizeUnit.MByte;
    case "mbyte":
      return SizeUnit.GByte;
    case "gbyte":
      return SizeUnit.TByte;
    case "tbyte":
      return SizeUnit.Unknown;
    default:
      return null;
  }
}

function sizeUnitToString(sizeUnit: SizeUnit): string {
  switch (sizeUnit) {
    case "byte":
      return "B";
    case "kbyte":
      return "K";
    case "mbyte":
      return "M";
    case "gbyte":
      return "G";
    case "tbyte":
      return "T";
    default:
      return "-";
  }
}

class Size {
  private data: SizeData;

  constructor(size: bigInt.BigInteger) {
    this.data = this.toData(size);
  }
  /**
   * convert to String
   */
  public toString(): string {
    const unit = sizeUnitToString(this.data.sizeUnit);
    const fixed = this.data.alignedSize.toFixed(1);
    const padded = fixed.padStart(5, " ");
    return `${padded}${unit}`;
  }

  /**
   * parse size string to data structure
   * @param size string representation of size
   * @return size data
   */
  private toData(size: bigInt.BigInteger): SizeData {
    function calcUnit(
      fileSize: bigInt.BigInteger,
      current: SizeUnit,
      decimal: number
    ): { current: SizeUnit; size: number } {
      if (bigInt.zero.leq(fileSize) && fileSize.lt(bigInt(1024))) {
        return { current, size: fileSize.toJSNumber() };
      }

      const unit = nextUnit(current);
      if (unit == null) {
        return { current, size: fileSize.toJSNumber() + decimal };
      } else {
        return calcUnit(fileSize.divide(bigInt(1024)), unit, fileSize.mod(bigInt(1024)).toJSNumber() / 1024);
      }
    }
    const parsedSize = bigInt(size);
    const data = calcUnit(parsedSize, SizeUnit.Byte, 0);

    return {
      sizeUnit: data.current,
      alignedSize: data.size,
      size: parsedSize,
    };
  }
}

export type Props = {
  size: bigInt.BigInteger;
};

const SizeNode = styled.pre`
  flex: 0 1 auto;
  padding: 0 ${props => props.theme.spaces.base};
  margin: 0;
  text-align: right;

  white-space: pre;
`;

export const Component: React.FC<Props> = prop => {
  const size = new Size(prop.size);

  return <SizeNode>{size.toString()}</SizeNode>;
};
