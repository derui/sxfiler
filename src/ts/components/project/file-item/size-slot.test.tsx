import { h } from "preact";
import { render } from "preact-render-to-string";
import { Component as T } from "./size-slot";

describe("Project", () => {
  describe("Node Item", () => {
    describe("Size slot", () => {
      const kb = BigInt(1024);
      it("should print zero size", () => {
        const tree = render(<T size={BigInt(0)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print byte size when size less than 1024", () => {
        const tree = render(<T size={BigInt(1023)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print kibibyte size when size between 1KiB and 1023KiB", () => {
        const tree = render(<T size={BigInt(1024) * BigInt(2)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print MiB size when size between 1MiB and 1023MiB", () => {
        const tree = render(<T size={kb * kb * BigInt(3)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print GiB size when size between 1GiB and 1023GiB", () => {
        const tree = render(<T size={kb * kb * kb * BigInt(5)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print TiB size when size between 1TiB and 1023TiB", () => {
        const tree = render(<T size={kb * kb * kb * kb * BigInt(5)} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print hyphen only when unit can not calculate", () => {
        const tree = render(<T size={kb * kb * kb * kb * kb} />);

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
