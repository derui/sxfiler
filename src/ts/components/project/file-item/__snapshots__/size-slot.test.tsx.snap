// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node Item Size slot should print GiB size when size between 1GiB and 1023GiB 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    5.0G
</pre>
`;

exports[`Project Node Item Size slot should print MiB size when size between 1MiB and 1023MiB 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    3.0M
</pre>
`;

exports[`Project Node Item Size slot should print TiB size when size between 1TiB and 1023TiB 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    5.0T
</pre>
`;

exports[`Project Node Item Size slot should print byte size when size less than 1024 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
  1023.0B
</pre>
`;

exports[`Project Node Item Size slot should print hyphen only when unit can not calculate 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    1.0-
</pre>
`;

exports[`Project Node Item Size slot should print kibibyte size when size between 1KiB and 1023KiB 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    2.0K
</pre>
`;

exports[`Project Node Item Size slot should print zero size 1`] = `
<pre
  className="sc-bdVaJa eztgmS"
>
    0.0B
</pre>
`;
